arwen/contexts/validator.go

```diff
@@ -28,16 +28,44 @@ func (validator *WASMValidator) verifyMemoryDeclaration(instance *wasmer.Instanc
 	return nil
 }
 
-func (validator *WASMValidator) verifyFunctionsNames(instance *wasmer.Instance) error {
+func (validator *WASMValidator) verifyFunctions(instance *wasmer.Instance) error {
 	for functionName := range instance.Exports {
 		if !validator.isValidFunctionName(functionName) {
 			return fmt.Errorf("%w: %s", arwen.ErrInvalidFunctionName, functionName)
 		}
+
+		if !validator.isVoidFunction(instance, functionName) {
+			return fmt.Errorf("%w: %s", arwen.ErrFunctionNonvoidSignature, functionName)
+		}
 	}
 
 	return nil
 }
 
+func (validator *WASMValidator) isVoidFunction(instance *wasmer.Instance, functionName string) bool {
+	inArity := validator.getInputArity(instance, functionName)
+	outArity := validator.getOutputArity(instance, functionName)
+
+	isVoid := (inArity == 0 && outArity == 0)
+	return isVoid
+}
+
+func (validator *WASMValidator) getInputArity(instance *wasmer.Instance, functionName string) int {
+	signature, ok := instance.Signatures[functionName]
+	if !ok {
+		return -1
```

Comment:
Changed these functions to return errors instead of `-1`.

Info: `functionName` is added to the `Signatures` map by reading the Wasmer exports directly, so if the WASM module does not export it explicitly at compilation time (in the `.export` file), it will not appear in `instance.Signatures`.